src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/MintermClassifier.cs

```diff
@@ -20,81 +20,124 @@ namespace System.Text.RegularExpressions.Symbolic
-        /// <param name="solver">The character set solver to use.</param>
-        public MintermClassifier(BDD[] minterms, CharSetSolver solver)
+        public MintermClassifier(BDD[] minterms)
         {
             Debug.Assert(minterms.Length > 0, "Requires at least");
 
+
             if (minterms.Length == 1)
             {
                 // With only a single minterm, the mapping is trivial: everything maps to it (ID 0).
-                // For ASCII, use an array containing all zeros.  For non-ASCII, use a BDD that maps everything to 0.
-                _ascii = AllAsciiIsZeroMintermArray;
-                _nonAscii = solver.ReplaceTrue(BDD.True, 0);
+                _lookup = Array.Empty<byte>();
+                _maxChar = -1;
                 return;
             }
 
-            // Create a multi-terminal BDD for mapping any character to its associated minterm.
-            BDD anyCharacterToMintermId = BDD.False;
-            for (int i = 0; i < minterms.Length; i++)
+            // attempt to save memory in common cases by allocating only up to the highest char code
+            for (int mintermId = 1; mintermId < minterms.Length; mintermId++)
             {
-                // Each supplied minterm BDD decides whether a given character maps to it or not.
-                // We need to combine all of those into a multi-terminal BDD that decides which
-                // minterm a character maps to.  To do that, we take each minterm BDD and replace
-                // its True result with the ID of the minterm, such that a character that would
-                // have returned True for that BDD now returns the minterm ID.
-                BDD charToTargetMintermId = solver.ReplaceTrue(minterms[i], i);
-
-                // Now union this BDD with the multi-terminal BDD we've built up thus far. Unioning
-                // is valid because every character belongs to exactly one minterm and thus will
-                // only map to an ID instead of False in exactly one of the input BDDs.
-                anyCharacterToMintermId = solver.Or(anyCharacterToMintermId, charToTargetMintermId);
+                _maxChar = Math.Max(_maxChar, (int)BDDRangeConverter.ToRanges(minterms[mintermId])[^1].Item2);
             }
-
-            // Now that we have our mapping that supports any input character, we want to optimize for
-            // ASCII inputs.  Rather than forcing every input ASCII character to consult the BDD at match
-            // time, we precompute a lookup table, where each ASCII character can be used to index into the
-            // array to determine the ID for its corresponding minterm.
-            var ascii = new int[128];
-            for (int i = 0; i < ascii.Length; i++)
+            // there is an opportunity to gain around 5% performance for allocating the
+            // full 64K, past a certain threshold where maxChar is already large.
+            // TODO: what should this threshold be?
```

Comment:
I guess this TODO can be removed? Or replaced with explanation of how this was chosen.

Reply:
This is currently just an arbitrary choice to allocate an extra 32kb for performance if 32kb is already allocated. Just needed to check if this is ok

Reply:
gotcha, you're expecting to remove TODO's

Reply:
Initially the idea was to make the next state transition store zero intermediate local variables, which would make it faster at the expense of the 64K allocation. In the case of re# it made a noticeable difference (`3.143 ms to 2.911 ms`)  since it's one of the very few operations at the very center of the match loop. I suppose it's not good to have mix of both conserving memory and this

Reply:
Thanks. But concretely here where did that cited 5% imprudent come from? eg where was "to make the next state transition store zero intermediate local variables" encoded?

Reply:
Ah this was not explained here, The full array allowed using an array lookup in place of the small input reader.
So turning this:
```csharp
char c = input[pos];
return c < (uint)lookup.Length ? lookup[c] : 0;
```
to 
```csharp 
return lookup[input[pos]];
```

Reply:
Yes, but the common case in both of those is essentially identical. Both have the same branches / bounds checks in the resulting assembly, it's just that it's explicit in the C# in the first and implicit / created by the JIT in the second. The only meaningful difference between them is where the failure branches to, whether returning 0 or throwing an exception.

Reply:
Oh, then you're right it is irrelevant here.